compiler/plc_ast/src/visitor.rs

```diff
@@ -0,0 +1,584 @@
+/// This module defines the `AstVisitor` trait and its associated macros.
+/// The `AstVisitor` trait provides a set of methods for traversing and visiting ASTs
+use crate::ast::AstNode;
+use crate::ast::*;
+use crate::control_statements::{AstControlStatement, ConditionalBlock, ReturnStatement};
+use crate::literals::AstLiteral;
+
+/// Macro that walks through all nodes in an iterator and applies the visitor's `walk` method to each node.
+macro_rules! visit_all_nodes {
+    ($visitor:expr, $iter:expr) => {
+        for node in $iter {
+            $visitor.visit(node);
+        }
+    };
+}
+
+/// Macro that walks through a list of nodes and applies the visitor's `walk` method to each node.
+macro_rules! visit_nodes {
+    ($visitor:expr, $($node:expr),*) => {
+        $(
+            $visitor.visit($node);
+        )*
+    };
+}
+
+pub trait Walker {
```

Comment:
Can you also give a quick documentation what the `Walker` trait is supposed to do and maybe how it differs from the `AstVisitor` trait? If I understood it correctly it's just an entry-point to various `visit_*` calls with some extra destructuring if needed?

Reply:
yes I'll add some additional documentation.
I looked at rust-lang's implementation. they say the following in their documentation:

https://github.com/rust-lang/rust/blob/master/compiler/rustc_ast/src/visit.rs#L1 
```
//! AST walker. Each overridden visit method has full control over what
//! happens with its node, it can do its own traversal of the node's children,
//! call `visit::walk_*` to apply the default traversal algorithm, or prevent
//! deeper traversal by doing nothing.
```
So if you override the default implementation of the visitor, you basically interrupted the visiting of your children. You got to make sure yourself, that the visiting continues. In oop languages you would just call `super.walk_*`, because what you usually want, is to just "hook into" the visiting at certain points (e.g. at assignments, or at If statements, etc.). So what rust says is, that if you implement the visit_X method in your concrete visitor, you should call the walk_X method to continue the traversal. So the key here is, to split the visiting (hooking in) from the walking (traversal) because in rust you cannot call the trait's default implementation, that you just overwrote.

I just picked a trait approach (`stmt.walk(...)`) instead of rust-lang's `walk_X` functions.

Reply:
done

Reply:
Makes sense, thanks for elaborating